### PR TITLE
Simple Carla controller

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This PR implements a simple controller for Carla.
It simply uses the navigation signal plus the PID controller. 

A success run without other vehicles on the street is shown below:

ezgif com-gif-maker (8)

### CODE REVIEW

File: alf/algorithms/handcrafted_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        """A naive hand-crafted policy for Carla environment.
+
+        Args:
+            observation (nested Tensor): input observation that is compatible
+                with observation_spec
+        Returns
+            Tensor: action computed based on the observation
+        """
+        nav = alf.nest.get_field(observation, 'observation.navigation')
+        goal = alf.nest.get_field(observation, 'observation.goal')
```
Comment (emailweixu) [inline alf/algorithms/handcrafted_algorithm.py:163]: perhaps add a comment about the data in nav to make the code more understandable.

File: alf/algorithms/handcrafted_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        waypoints = nav
+        if waypoints.shape[-1] > 1:
+            wp_vector = waypoints[:, 1]
+        else:
+            wp_vector = waypoints[:, 0]
```
Comment (emailweixu) [inline alf/algorithms/handcrafted_algorithm.py:168]: The shape of waypoints is [B, k, 3], where B is the batch size, k is the number of way points.
Should the if statement be `if waypoints.shape[1] > 1`?

File: alf/algorithms/handcrafted_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import alf
+from alf.algorithms.config import TrainerConfig
+from alf.algorithms.off_policy_algorithm import OffPolicyAlgorithm
+from alf.data_structures import AlgStep, Experience, LossInfo, TimeStep
+from alf.tensor_specs import BoundedTensorSpec
+
+
+@gin.configurable
+class HandcraftedAlgorithm(OffPolicyAlgorithm):
+    """An Algorithm with handcrafted computational logic.
```
Comment (emailweixu) [inline alf/algorithms/handcrafted_algorithm.py]: Comment that a concrete algorithm should subclass from this and implement _policy_func

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/algorithms/handcrafted_algorithm.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/handcrafted_algorithm.py]: Comments added

File: alf/algorithms/handcrafted_algorithm.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/handcrafted_algorithm.py:168]: Bug fixed.

File: alf/algorithms/handcrafted_algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/handcrafted_algorithm.py:163]: Added comments on the data in nav, goal etc to provide more context to the code.

File: alf/algorithms/handcrafted_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        # reverse: values greater than 0.5 corrsponding to going backward.
+
+        action = torch.zeros(waypoints.shape[0], self._action_spec.shape[0])
+
+        distance_to_goal = torch.norm(goal, dim=1)
+
+        # here we adjust the speed based on the distance to goal
+        action[distance_to_goal > self._distance_to_decelerate, 0] = 1
+        # logical_and is introduced in torch 1.5.0, use * instead
+        ind = (distance_to_goal > self._distance_to_stop) * (
```
Comment (emailweixu) [inline alf/algorithms/handcrafted_algorithm.py]: FYI: & can be used

**Review by emailweixu — DISMISSED** [review]

File: alf/algorithms/handcrafted_algorithm.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/algorithms/handcrafted_algorithm.py]: Good suggestion. I have removed the comment and changed to &

**Review by emailweixu — APPROVED** [review]

**Pull request merged**